shm_wrapper/shm_wrapper.h

```diff
@@ -16,122 +16,225 @@
+ *    - process_t process: the calling process (will error if not EXECUTOR or NET_HANDLER)
+ *    - uint32_t params_to_sub: bitmap representing params to subscribe to (nonexistent params should have corresponding bits set to 0)
+ * Returns 0 on success, -1 on failure (unrecognized process, or device is not connect in shm)
+ */
+int place_sub_request (uint64_t dev_uid, process_t process, uint32_t params_to_sub);
+
+/*
+ * Get current subscription requests for all devices. Should only be called by dev_handler
+ * Arguments:
+ *    - uint32_t *sub_map: bitwise OR of the executor and net_handler sub_maps that will be put into this provided buffer
+ * No return value.
+ */
+void get_sub_requests (uint32_t *sub_map);
+
+/*
+ * Reads the specified robot description field. Blocks on the robot description semaphore.
+ * Arguments:
+ *    - field: one of the robot_desc_val_t's defined above to read from
+ * Returns one of the robot_desc_val_t's defined in runtime_util that is the current value of the requested field.
+ */
+robot_desc_val_t robot_desc_read (robot_desc_field_t field);
+
+/*
+ * Writes the specified value into the specified field. Blocks on the robot description semaphore.
+ * Arguments:
+ *    - robot_desc_field_t field: one of the robot_desc_val_t's defined above to write val to
+ *    - robot_desc_val_t val: one of the robot_desc_vals defined in runtime_util.c to write to the specified field
+ * No return value.
+ */
+void robot_desc_write (robot_desc_field_t field, robot_desc_val_t val);
+
+/*
+ * Reads current state of the gamepad to the provided pointers. 
+ * Blocks on both the gamepad semaphore and device description semaphore (to check if gamepad connected).
+ * Arguments:
+ *    - uint32_t pressed_buttons: pointer to 32-bit bitmap to which the current button bitmap state will be read into
+ *    - float *joystick_vals: array of 4 floats to which the current joystick states will be read into
+ * Returns 0 on success, -1 on failure (if gamepad is not connected)
+ */
+int gamepad_read (uint32_t *pressed_buttons, float *joystick_vals);
+
+/*
+ * This function writes the given state of the gamepad to shared memory.
+ * Blocks on both the gamepad semaphore and device description semaphore (to check if gamepad connected).
+ * Arguments:
+ *    - uint32_t pressed_buttons: a 32-bit bitmap that corresponds to which buttons are currently pressed
+            (only the first 17 bits used, since there are 17 buttons)
+ *    - float *joystick_vals: array of 4 floats that contain the values to write to the joystick
+ * Returns 0 on success, -1 on failuire (if gamepad is not connected)
+ */
+int gamepad_write (uint32_t pressed_buttons, float *joystick_vals);
+
+/*
+ * Should be called from all processes that want to know current state of the command map (i.e. device handler)
+ * Blocks on the command bitmap semaphore for obvious reasons
+ * Arguments:
+ *    - uint32_t *bitmap: pointer to array of 17 32-bit integers to copy the bitmap into
```

Comment:
Array of thirty-three uint32_t instead of seventeen. Maybe for clarity add that the 0th index is a bitmap indicating which cmd_maps got changed and index "i + 1" is the actual command bitmap for device i

Reply:
Same comment as above for `get_sub_requests` with `void get_cmd_map (uint32_t bitmap[])`

Reply:
I just reference the README, since I explain in more detail how the command bitmap works in there~~
👌